solver/cvrp.py

```diff
@@ -407,66 +393,54 @@ def parse_solution_nl(self, tolerance=1e-6) -> None:
-            """Compute the objective given a solution."""
-            total_cost = 0
-            assert len(solution) == num_vehicles
-
-            # Compute total cost for the solution.
-            for r in solution:
-                if len(r) == 0:
-                    continue
-
-                for index, location in enumerate([0, *r[:-1]]):
-                    total_cost += self._costs[all_locations[location], all_locations[r[index]]]
-
-                total_cost += self._costs[
-                    all_locations[r[-1]], all_locations[0]  # Go back to depot
-                ]
+        assert len(solution) == num_vehicles
 
-            return total_cost
+        for r in solution:
+            # If route has no locations the vehicle never left the depot or
+            # if demand exceeds capacity.
+            if len(r) == 0 or demand[r].sum() > max_capacity:
+                return False
 
-        def check_feasibility(solution):
-            """Check whether the given solution is feasible"""
-            assert len(solution) == num_vehicles
-
-            for r in solution:
-                if len(r) == 0:  # If route has no locations the vehicle never left the depot.
-                    return False
-
-                if demand[r].sum() > max_capacity:  # If demand exceeds capacity
-                    return False
-
-            return True
+        return True
 
+    def _get_solution(self, tolerance=1e-6):
+        """Extract solution and check feasibility"""
+        model = self._optimization["nl"]
         num_states = model.states.size()
-        solutions = []
         for i in range(num_states):
-            # extract the solution from the route NL variables
-            solution = [[int(v) + 1 for v in route.state(i)] for route in routes]
+            # extract the solution
+            decision = next(model.iter_decisions())
+            solution_candidate = [[int(v) + 1 for v in route.state(i)] for route in decision.iter_successors()]
+            if not solution_candidate: continue
 
             solver_objective = model.objective.state(i)
-            assert abs(solver_objective - recompute_objective(solution)) < tolerance
+            assert abs(solver_objective - self._recompute_objective(solution=solution_candidate)) < tolerance
 
             solver_feasibility = True
             for c in model.iter_constraints():
                 if c.state(i) < 0.5:
                     solver_feasibility = False
 
-            # Check feasibility and if feasible, add to solutions list
-            assert solver_feasibility == check_feasibility(solution)
-            if not solver_feasibility:
-                print(f"Sample {i} is infeasible")
-            else:
-                solutions.append(solution)
+            # Check feasibility and if feasible, return
+            assert solver_feasibility == self._check_feasibility(solution=solution_candidate)
+            if solver_feasibility:
```

Comment:
This could probably be shortened to
```suggestion
            if solver_feasibility and self._check_feasibility(solution_candidate):
```
with the only difference being that if the feasibility-check is False, the next solution is checked instead of an assertion error being raised, or actually raise an exception if we think it might ever fail.
```suggestion
            if solver_feasibility != self._check_feasibility(solution_candidate):
                raise ValueError("Solution failed feasibility comparison check.")
            if solver_feasibility:
```
But I think the latter should never occur.

Reply:
Also, as a side note, we should probably not use `ValueError` here and below, instead adding something like a `SolutionNotFound` exception.